stl/inc/ranges

```diff
@@ -6821,6 +6821,682 @@ namespace ranges {
+            requires(sized_range<_ViewTypes>&&...) {
+            using _Common_size_type          = common_type_t<range_size_t<_ViewTypes>...>;
+            using _Common_unsigned_size_type = _Make_unsigned_like_t<_Common_size_type>;
+
+            auto _Smallest_view_size_value = (numeric_limits<_Common_unsigned_size_type>::max)();
+
+            _Zip::_Tuple_for_each(
+                [&_Smallest_view_size_value]<class _View>(_View& _Curr_view) {
+                    const auto _Curr_view_size_value =
+                        static_cast<_Common_unsigned_size_type>(_RANGES size(_Curr_view));
+
+                    if (_Curr_view_size_value < _Smallest_view_size_value)
+                        _Smallest_view_size_value = _Curr_view_size_value;
+                },
+                _Ranges);
+
+            return _Smallest_view_size_value;
+        }
+
+        _NODISCARD constexpr auto size() const
+            noexcept(noexcept(_Zip::_Tuple_for_each(_RANGES size, _Ranges))) // strengthened
+            requires(sized_range<const _ViewTypes>&&...) {
+            using _Common_size_type          = common_type_t<range_size_t<_ViewTypes>...>;
+            using _Common_unsigned_size_type = _Make_unsigned_like_t<_Common_size_type>;
+
+            auto _Smallest_view_size_value = (numeric_limits<_Common_unsigned_size_type>::max)();
+
+            _Zip::_Tuple_for_each(
+                [&_Smallest_view_size_value]<class _View>(const _View& _Curr_view) {
+                    const auto _Curr_view_size_value =
+                        static_cast<_Common_unsigned_size_type>(_RANGES size(_Curr_view));
+
+                    if (_Curr_view_size_value < _Smallest_view_size_value)
+                        _Smallest_view_size_value = _Curr_view_size_value;
+                },
+                _Ranges);
+
+            return _Smallest_view_size_value;
+        }
+
+    private:
+        _NODISCARD constexpr auto _Get_signed_size() noexcept(noexcept(size())) {
+            const auto _Possibly_unsigned_size = size();
+
+            if constexpr (is_unsigned_v<decltype(_Possibly_unsigned_size)>) {
+                return static_cast<_Make_signed_like_t<decltype(_Possibly_unsigned_size)>>(_Possibly_unsigned_size);
+            } else {
+                return _Possibly_unsigned_size;
+            }
+        }
+
+        _NODISCARD constexpr auto _Get_signed_size() const noexcept(noexcept(size())) {
+            const auto _Possibly_unsigned_size = size();
+
+            if constexpr (is_unsigned_v<decltype(_Possibly_unsigned_size)>) {
+                return static_cast<_Make_signed_like_t<decltype(_Possibly_unsigned_size)>>(_Possibly_unsigned_size);
+            } else {
+                return _Possibly_unsigned_size;
+            }
+        }
+
+        template <bool _IteratorConst>
+        _NODISCARD static constexpr const auto& _Get_iterator_tuple(const _Iterator<_IteratorConst>& _Itr) noexcept {

```

Comment:
That's odd. What you are saying should work in my tests, and removing `_Get_iterator_tuple` and just accessing the `_Current` member of the iterators works on Clang, but it fails to compile on the MSVC with error C2248. I'm inclined to believe that this is some form of compiler bug.

I've been reviewing the standard regarding this. Correct me if I am wrong, but my understanding is that if a nested class `InnerClassA` names its enclosing class `OuterClass` as a friend (as by `friend OuterClass`), then any member of `OuterClass` should have the same access rights as any member of `InnerClassA`. Thus, since nested classes are members, any additional nested class within `OuterClass`, such as `InnerClassB`, should be able to access the members of `InnerClassA`.

I believe this is what you are getting at here, and if I am interpreting this right, then this should work. Despite that, the MSVC is throwing error C2248 and stating that `_Sentinel` cannot access the private members of `_Iterator`. To work around this, I am going to be leaving `_Get_iterator_tuple()` intact for now, since this method works with both Clang and the MSVC. Let me know if you can think of any alternatives.

Reply:
Alright, having done the investigation, looks like this is a bug in clang, not in MSVC (see attached godbolt links) - `operator==` is a friend of `_Sentinel`, not a member, and friendship isn't transitive, so it should not have access to `_Iterator`'s private members. LLVM bug opened at LLVM-57723

* clang incorrectly treating a `friend` function as a member - <https://gcc.godbolt.org/z/rocdhvnYr>
* clang correctly disallowing transitive friendship - <https://gcc.godbolt.org/z/db1vnvse6>